File: src/com/cattailsw/nanidroid/GhostMgr.java
package com.cattailsw.nanidroid;

import android.content.Context;
import java.util.List;

import com.cattailsw.nanidroid.util.PrefUtil;
import com.cattailsw.nanidroid.util.NarUtil;
import java.io.File;

public class GhostMgr {
    private static final String TAG = "GhostMgr";
    private static final String PREF_LAST_RUN_GHOST = "lastrunghost";
    Context mCtx;

    List<InfoOnlyGhost> iglist = null;

    public GhostMgr(Context ctx) {
	mCtx = ctx.getApplicationContext();
	iglist = DirList.parseDataDir(mCtx);
    }

    public int getGhostId(String name){
	int id = 0;
	for ( InfoOnlyGhost g: iglist ){
	    if ( g.getGhostDirName().equalsIgnoreCase(name) )
		return id;
	    id++;
	}
	return -1;
    }

    public boolean hasSameGhostId(String id){
	if ( iglist == null || iglist.size() == 0 )
	    return false;

	return (getGhostId(id) != -1);
    }

    public String getGhostPath(int id){
	return iglist.get(id).getGhostPath();
    }

    public Ghost createGhost(String name){
	int id = getGhostId(name);
	if ( id == -1 )
	    return null;

	return new Ghost(getGhostPath(id));
    }

    public String getLastRunGhostId(){
	if ( PrefUtil.hasKey(mCtx, PREF_LAST_RUN_GHOST))
	    return PrefUtil.getKeyValue(mCtx, PREF_LAST_RUN_GHOST);

	return null;
    }

    public void setLastRunGhost(Ghost g){
	String gid = g.getGhostDirName();

	PrefUtil.setKey(mCtx, PREF_LAST_RUN_GHOST, gid);
    }

    public String installFirstGhost(String gid, String narPath){
    	return installGhost(gid, narPath, true);
    }
    
    public String installGhost(String gid, String narPath){
    	return installGhost(gid, narPath, false);
    }
    
    public String installGhost(String ghostId, String narPath, boolean usegid){
	File dataDir = new File(mCtx.getExternalFilesDir(null), "ghost");
	boolean success = NarUtil.readNarArchive(narPath, dataDir.getAbsolutePath(), 
				usegid?ghostId:null);
	if ( success == false)
		return null;
	
	refreshGhost();
	int gid =  getGhostId( ghostId);
	if ( gid == -1 ) return null;
	String path = getGhostPath(gid);
	return path;
    }

    public void refreshGhost(){
	iglist = DirList.parseDataDir(mCtx);
    }

    public String[] getGnames(){
	if ( iglist == null || iglist.size() == 0 )
	    return null;

	String []ret = new String[iglist.size()];
	int i =0;
	for ( InfoOnlyGhost g: iglist ) {
	    ret[i] = g.getGhostDirName();
	    i++;
	}
	return ret;
    }
    
    public int getGhostCount(){
    	return (iglist == null)?0:iglist.size();
    }
}

File: src/main/kotlin/com/cattailsw/nanidroid/GhostMgr.kt
package com.cattailsw.nanidroid

import android.content.Context
import com.cattailsw.nanidroid.install.ArchiveInstallFailure
import com.cattailsw.nanidroid.install.ArchiveInstallResult
import com.cattailsw.nanidroid.install.NarTransactionalInstaller
import com.cattailsw.nanidroid.util.PrefUtil
import java.io.File

internal fun shouldInstallBundledGhost(
    usableGhostCount: Int,
    storageEntries: Array<out File>,
): Boolean = usableGhostCount == 0 && storageEntries.all { entry ->
    entry.isDirectory && entry.name == ".nanidroid-install-staging"
}

/** Kotlin owner for ghost discovery, selection, and fresh installation. */
internal class GhostMgr(
    ctx: Context,
    private val ghostRuntime: GhostRuntime,
) {
    private val context = ctx.applicationContext
    private var ghosts: List<InfoOnlyGhost>? = loadGhosts()

    fun getGhostId(name: String): Int {
        ghosts?.forEachIndexed { index, ghost ->
            if (ghost.getGhostDirName().equals(name, ignoreCase = true)) return index
        }
        return -1
    }

    fun getGhostPath(id: Int): String = ghosts!![id].getGhostPath()

    internal fun createGhost(name: String): ReservedGhost? {
        val id = getGhostId(name)
        if (id == -1) return null
        val root = File(getGhostPath(id)).canonicalFile
        return ghostRuntime.reuseActiveGhost(root.name, root) ?: run {
            val construction = ghostRuntime.beginGhostConstruction(root.name, root)
            try {
                construction.bind(Ghost(root.path, context))
            } catch (error: Exception) {
                construction.failConstruction()
                throw error
            } catch (error: LinkageError) {
                construction.failConstruction()
                throw error
            }
        }
    }

    fun getLastRunGhostId(): String? =
        if (PrefUtil.hasKey(context, PREF_LAST_RUN_GHOST)) {
            PrefUtil.getKeyValue(context, PREF_LAST_RUN_GHOST)
        } else {
            null
        }

    fun setLastRunGhost(ghost: Ghost) {
        PrefUtil.setKey(context, PREF_LAST_RUN_GHOST, ghost.getGhostDirName())
    }

    fun installFirstGhost(
        gid: String,
        narPath: String,
        isCancelled: () -> Boolean,
    ): ArchiveInstallResult = installGhost(gid, narPath, true, isCancelled)

    fun installGhost(
        ghostId: String,
        narPath: String,
        usegid: Boolean,
        isCancelled: () -> Boolean,
    ): ArchiveInstallResult {
        val externalFiles = context.getExternalFilesDir(null)
        if (externalFiles == null) {
            return ArchiveInstallResult.Failed(
                "Nanidroid cannot access the selected ghost archive or storage.",
                ArchiveInstallFailure.StorageUnavailable,
            )
        }
        val dataDir = File(externalFiles, "ghost")
        if ((!dataDir.exists() && !dataDir.mkdirs()) || !dataDir.isDirectory) {
            return ArchiveInstallResult.Failed(
                "Nanidroid cannot prepare its ghost storage.",
                ArchiveInstallFailure.StorageUnavailable,
            )
        }
        val installed = NarTransactionalInstaller.install(
            File(narPath), dataDir, if (usegid) ghostId else null, isCancelled
        )
        if (installed !is ArchiveInstallResult.Installed) {
            return installed
        }
        refreshGhost()
        val id = getGhostId(installed.targetId!!)
        if (id == -1) {
            return ArchiveInstallResult.Failed(
                "The installed archive does not contain a usable ghost.",
                ArchiveInstallFailure.InvalidArchive,
            )
        }
        return ArchiveInstallResult.Installed(getGhostPath(id), installed.targetId)
    }

    fun refreshGhost() {
        ghosts = loadGhosts()
    }

    private fun loadGhosts(): List<InfoOnlyGhost>? = DirList.parseDataDir(context)

    fun getGnames(): Array<String>? =
        ghosts?.takeIf { it.isNotEmpty() }?.map { it.getGhostDirName() }?.toTypedArray()

    fun getGhostCount(): Int = ghosts?.size ?: 0

    fun shouldInstallFirstGhost(): Boolean {
        val externalFiles = context.getExternalFilesDir(null) ?: return false
        val storageRoot = File(externalFiles, "ghost")
        return shouldInstallBundledGhost(getGhostCount(), storageRoot.listFiles().orEmpty())
    }

    fun getGhostReadMe(ghostId: String): File =
        File(getGhostPath(getGhostId(ghostId)), "readme.txt")

    fun getGhostSakuraName(id: String): String? =
        getGhostId(id).takeIf { it != -1 }?.let { ghosts!![it].getSakuraName() }

    fun getGhostDispName(id: String): String? =
        getGhostId(id).takeIf { it != -1 }?.let { ghosts!![it].getGhostName() }

    fun getGDispNames(): Array<String?>? =
        ghosts?.takeIf { it.isNotEmpty() }?.map { it.getGhostName() }?.toTypedArray()

    fun getGhostPath(id: String): String = getGhostPath(getGhostId(id))

    private companion object {
        const val PREF_LAST_RUN_GHOST = "lastrunghost"
    }
}
